PathPlanning/BreadthFirstSearch/breadth_first_search.py

```diff
@@ -0,0 +1,257 @@
+"""
+
+Breadth-First grid planning
+
+author: Erwin Lejeune (@spida_rwin)
+
+See Wikipedia article (https://en.wikipedia.org/wiki/Breadth-first_search)
+
+"""
+
+import math
+
+import matplotlib.pyplot as plt
+
+show_animation = True
+
+
+class BreadthFirstSearchPlanner:
+
+    def __init__(self, ox, oy, reso, rr):
+        """
+        Initialize grid map for bfs planning
+
+        ox: x position list of Obstacles [m]
+        oy: y position list of Obstacles [m]
+        reso: grid resolution [m]
+        rr: robot radius[m]
+        """
+
+        self.reso = reso
+        self.rr = rr
+        self.calc_obstacle_map(ox, oy)
+        self.motion = self.get_motion_model()
+
+    class Node:
+        def __init__(self, x, y, cost, pind, parent):
+            self.x = x  # index of grid
+            self.y = y  # index of grid
+            self.cost = cost
+            self.pind = pind
+            self.parent = parent
+
+        def __str__(self):
+            return str(self.x) + "," + str(self.y) + "," + str(
+                self.cost) + "," + str(self.pind)
+
+    def planning(self, sx, sy, gx, gy):
+        """
+        Breadth First search based planning
+
+        input:
+            sx: start x position [m]
+            sy: start y position [m]
+            gx: goal x position [m]
+            gy: goal y position [m]
+
+        output:
+            rx: x position list of the final path
+            ry: y position list of the final path
+        """
+
+        nstart = self.Node(self.calc_xyindex(sx, self.minx),
+                           self.calc_xyindex(sy, self.miny), 0.0, -1, None)
+        ngoal = self.Node(self.calc_xyindex(gx, self.minx),
+                          self.calc_xyindex(gy, self.miny), 0.0, -1, None)
+
+        open_set, closed_set = dict(), dict()
+        open_set[self.calc_grid_index(nstart)] = nstart
+
+        while 1:
+            if len(open_set) == 0:
+                print("Open set is empty..")
+                break
+
+            current = open_set.pop(list(open_set.keys())[0])
```

Comment:
I'm wander that it is correct for breadth-First search? The key of open_set is not ordered by insertion time, I think it is order by hash number. So, the first element is not oldest open set data. What do you think? I think we need to have a queue to store a key in order to manage the insertion time and it should be used.

Reply:
I think since in Python 3.6, dictionaries were "unofficially" ordered and guaranteed insertion order, and since Python 3.7 they're officially ordered in insertion order, so this works starting Python 3.6.
I don't know about prior versions, haven't tested. They do come out in insertion order for me though.

Reply:
Oh, I'm sorry you are right. I miss understand it as another language. This is the reference for it: https://stackoverflow.com/questions/5629023/the-order-of-keys-in-dictionaries